alua.c

```diff
@@ -483,10 +519,28 @@ int tcmu_transition_tgt_port_grp(struct tgt_port_grp *group, uint8_t new_state,
 	return SAM_STAT_GOOD;
 }
 
+static int tcmu_report_state(struct tcmu_device *dev,
+			     struct tgt_port_grp *group)
+{
+	struct tcmur_handler *rhandler = tcmu_get_runner_handler(dev);
+	int ret;
+
+	/* TODO: For ESX return remote ports */
+
+	if (!rhandler->has_lock)
+		return group->state;
+
+	ret = rhandler->has_lock(dev);
+	if (ret <= 0) {
+		return ALUA_ACCESS_STATE_STANDBY;
+	} else {
+		return ALUA_ACCESS_STATE_OPTIMIZED;
+	}
+}
+
 int tcmu_emulate_report_tgt_port_grps(struct tcmu_device *dev,
 				      struct list_head *group_list,
-				      struct tcmulib_cmd *cmd,
-				      tcmu_report_state_fn_t *report_fn)
+				      struct tcmulib_cmd *cmd)
```

Comment:
> The alua.c code is currently in libtcmu and was able to support non tcmu-runner based daemons. That is why we passed in the report/transition callouts. So, you need to either pass the lock callouts in or just move the alua.c code from libtcmu to tcmu-runner.

By design, yes you are right but in reality, I think it is okay to move fast, leaving this to be settled down later. If anyone finds alua mechanism too fixed to change for him, he should make a pr. Besides, at least he can implement .handle_cmd to take over and put all the ALUA logic in its own deamon.   Anyway, I guess there is no serieous libtcmu user that implement its own target daemon. Most users will simply run tcmu :)